### ISSUE TITLE
Accessibility

### ISSUE THREAD

**Issue by aaronoppenheimer** [Gig-o-Matic/GO3#778]
A band reached out and said that they have a blind member, and what do we think about improving accessibility? That would be great! I started playing with the screenreader built into the mac, and immediately ran into problems: although the documentation all insists that things like \<h1\> tags should be recognized and the text should be read, it isn't. I can only get actual buttons/controls to be recognized, all other text is ignored.

**Comment by kjcole** [comment]
> A band reached out and said that they have a blind member, and what do we think about improving accessibility? That would be great! I started playing with the screenreader built into the mac, and immediately ran into problems: although the documentation all insists that things like `<h1>` tags should be recognized and the text should be read, it isn't. I can only get actual buttons/controls to be recognized, all other text is ignored.

@aaronoppenheimer  Can you point to a specific example page -- or example page type -- where you're not hearing something? Is it a band description page? A gig listing?

**Comment by aaronoppenheimer** [comment]
The gig-o as released doesn't have any `<h1>` tags to show the problem, I've only seen it in my development environment. I'll update the templates to show the problem.

**Comment by kjcole** [comment]
> The gig-o as released doesn't have any `<h1>` tags to show the problem, I've only seen it in my development environment. I'll update the templates to show the problem.

Ah! I was wondering why I couldn't find any. Also, what browsers have you tried? Do all fail to pass the `<h1>` on to the screen reader?

**Comment by aaronoppenheimer** [comment]
OK - I pushed a change so that the "agenda" page has the title wrapped in `<h1>`.

I'm using "voiceover" on my mac, which is the built-in screenreader. I see different behavior with different browsers, and unsurprisingly Safari seems to act the best and recognize all of the navigation menus - but none of them read the header text.

I don't have access to a windows-based screenreader.

A

**Comment by uipoptart** [comment]
Hey @aaronoppenheimer,
Adding semantic markup like headings to your HTML is a great way to start building screen reader support. VoiceOver can be quite unintuitive to jump into. I am assuming that you are navigating with VoiceOver using the tab key or the VO key (this can be configured but it is usually Control+Option). When you navigate that way you'll stop on interactive elements like links and buttons. To access the heading elements you can use the Rotor (VO + U). The Rotor lets you navigate by element type, like heading. You use the left right arrows once in the Rotor to get to the list of headings to read. [Deque University has a good writeup on the VoiceOver shortcuts and how to use them.](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts#vo-mac-the-rotor) When I test on the schedule page the heading element is there and read as expected.
A couple additional suggestions-
The items in the green navigation bar like Gigs and Bands could benefit from more contrast. This would help visually impaired folks read the navigation items. The [WebAIM contrast checker](https://webaim.org/resources/contrastchecker/?fcolor=FFFFFFcc&bcolor=008000) is an easy way to see if the colors and font styles you've chosen pass accessibility recommendations.
For your Help page, one way to make it more accessible in addition to  adding an `h1` would be to build out [page structure](https://www.w3.org/WAI/tutorials/page-structure/headings/#organize-passages-of-text) with by wraping the questions in `h2`s. Because the Rotor will read each one off, I would suggest including the entire question in each heading. For the last question, a [helpful link text](https://webaim.org/techniques/hypertext/link_text) would be the purpose of the link, in this case to email the administrator. 
```<h1>Help</h1>
<h2>How do I join a band?</h2>
<p>Click your name in the upper-right of this window, and select "My Profile." In the "Bands" section, click "Add Band" and select the band of your choice! You'll be a pending member until the administrator for that band confirms your membership. You'll get an email about it.</p>
<h2>How do I  change my section?</h2>
<p>Click your name in the upper-right of this window, and select "My Profile." In the "Bands" section, select your section from the pop-up next to the band. Note that some bands aren't organized by section, so there might not be anything to select. So relax and enjoy the anarchy.</p>
<h2>How do I  make a suggestion?</h2>
<p>We want to [hear about it](mailto:superuser@gig-o-matic.com). Also, you can check the [issues list in github](https://github.com/gig-o-matic/GO3/issues?state=open)—what you're wishing for may already be on it.</p>
<h2>How do I contact the administrator?</h2>
<p>[Email our administrator](superuser@gig-o-matic.com)</p>
```

Hopefully this is helpful. If you'd like more insight I am happy to help, this is a cause near and dear to my heart and I appreciate other developers taking the time for it.
Grey

**Comment by kjcole** [comment]
@uipoptart Lots of good stuff there! (I'm on a Linux box, so all of the Apple Voice Over stuff is a mystery to me.) I'm including text from a message I sent via e-mail to @aaronoppenheimer on August 25, before entering the fray here, just so that it's in the thread and future contributors can find it.

> Making sure that gig creators follow accessibility rules -- or munging their submissions on the fly post-submit -- may be a bit challenging.
>
> The quick'n'dirty place to start is to run pages through the WAVE from WebAIM:
>
> https://wave.webaim.org/
>
> For some pages, it's nearly impossible to get a "perfect" score, but usually, there's lots of room for improvement. Also look at the ARIA attributes:
>
> https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes  
> https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA

**Comment by dianepanagiotopoulos** [comment]
@uipoptart @kjcole I've been working on some accessibility remediation for websites at my day job, and am familiar with a lot of the tools used to help diagnose common issues. My plan is to start with WAVE as a basic overview tool and then get more specific with screenreaders after the low-hanging fruit is picked off.

https://wave.webaim.org/

**Comment by allieorocks22** [comment]
>  I can only get actual buttons/controls to be recognized, all other text is ignored.

I think you might need an `aria-label` or an `aria-labelledby` to give the screen reader more direction. I'll put a couple examples below: 

Search Button with an Icon Only with 'aria label' :

html
`<button aria-label="Search">
  <svg></svg>
</button>
`

H1 html
`<h1 aria-label="Main Page Title">Welcome to Our Site</h1>`


I'm less of a programmer / coder than others in this thread (I'm a digital PM by trade) so I defer to the experts on whether my comment needs more context or detail. I hope this helps!

**Comment by dianepanagiotopoulos** [comment]
@aaronoppenheimer Before I get too far into the weeds on this, this is what I've been doing: https://github.com/dianepanagiotopoulos/GO3/tree/accessibility-remediation

Some of these changes are pretty straightforward (changing `div` elements to `h1`, or fixing broken labels), but some of them are a little more stylistic. Specifically, several of the errors are around low contrast between background color and text color. I've used [this contrast tool](https://webaim.org/resources/contrastchecker/) to find colors that are similar to what you already have, but that satisfy the minimum contrast requirement, in my changes. If you have any specific stylistic desires in terms of colors, please let me know, or I will just keep trucking on until it's time to submit this as a PR.

**Comment by aaronoppenheimer** [comment]
I'm not married to any of the colors, so changes within the same family (ie 'some kind of blue') are fine with me. The only exception is the navbar.dev style, which should keep the red color as the background of the menubar when running in debug mode. Otherwise I get confused when comparing the dev version to the deployed version!

I now understand the "rotor" business in VoiceOver and can make a little more sense of what I'm looking at (and hearing).

THANKS!

**Comment by aaronoppenheimer** [comment]
I am going to work on the 'agenda' page.

**Comment by acivilateJohnC** [comment]
Hi, I see some good suggestions in here. Thanks to @uipoptart for the Voiceover Rotor resource; an additional resource is  [Apple's Rotor support page](https://support.apple.com/guide/voiceover/voiceover-rotor-mchlp2719/mac) . I find the Rotor, and WAVE, helpful to give me a quick overview of what a screen reader can “see” on a web page; I then compare that with what I (as a sighted reader) can see, to identify what changes to make so a screen reader user can have an equivalent experience. 

For example, on the “Schedule” page, under “Form Controls”, there is a long list of “menu pop up collapsed button” (this band I am in has many gigs). As a sighted user, I can tell which button to use to respond to which gig; it's more difficult for a screen reader user to determine which button for which gig. This could be addressed as suggested by @allieorocks22  , adding an `aria-label` to each of those buttons, which would say something like `aria-label=“respond for gig {name of gig}”. `For example, I see in the branch from @dianepanagiotopoulos an `aria-label='{% trans "Comment on your status" %}'` - it may need to be something like `aria-label='{% trans "Comment on your status for {name of gig} " %}' `

Note that `aria-labeledby` is used when you have a separate HTML `label` tag, which shows and visually differentiates the label for the form element. I can’t tell if you are using the HTML `label` in gig-o, so you probably want to use `aria-label` instead. 

Another example from the “Schedule” page, under Links, there is a list of gig names, but the date/time is not included. The sighted reader can see the date/time next to the name of the gig; it’s going to be harder for a screen reader user to identify when a gig is, if only presented with the link text. (Consider if a band has a gig on multiple different days, but otherwise the description is identical.) This could be addressed by adding an `aria-label` to each of those links, which would say something like `aria-label=“on {date and time of gig} {name of gig}”.`  

I completely agree with the comment "Making sure that gig creators follow accessibility rules -- or munging their submissions on the fly post-submit -- may be a bit challenging." 

Sorry that I'm not more familiar with your development platform and code base. I would be happy to try to answer any specific questions.

**Comment by jcrisp78** [comment]
oops - the previous comment by @acivilateJohnC was by me, but  I accidentally used an old (and incorrect) github login for a previous employer...that may look a bit odd. Sorry...

**Comment by dianepanagiotopoulos** [comment]
@jcrisp78 That's a great suggestion about putting the gig name in the labels! Sorry to all that my progress on this has been slow, I'm mostly only able to work on it a couple of evenings a week, and last time I was catching up on a lot of Django paradigms.

**Comment by aaronoppenheimer** [comment]
I have started playing with doing that. Are there any good examples you all can point to of sites that do this kind of label well?

**Comment by dianepanagiotopoulos** [comment]
I've been working on the home page, gig create/edit, and tonight the member edit page. Will submit PR later tonight to ensure changes between multiple people stay in sync.

@aaronoppenheimer  - I have changes in my branch already on the agenda page, I hope this doesn't conflict with anything you're doing!
